### PR TITLE
Add support for `for await` loops

### PR DESCRIPTION
**Pull request by eholk**
This adds support for `for await` loops. This includes parsing, desugaring in AST->HIR lowering, and adding some support functions to the library.

Given a loop like:
```rust
for await i in iter {
    ...
}
```
this is desugared to something like:
```rust 
let mut iter = iter.into_async_iter();
while let Some(i) = loop {
    match core::pin::Pin::new(&mut iter).poll_next(cx) {
        Poll::Ready(i) => break i,
        Poll::Pending => yield,
    }
} {
    ...
}
```

This PR also adds a basic `IntoAsyncIterator` trait. This is partly for symmetry with the way `Iterator` and `IntoIterator` work. The other reason is that for async iterators it's helpful to have a place apart from the data structure being iterated over to store state. `IntoAsyncIterator` gives us a good place to do this.

I've gated this feature behind `async_for_loop` and opened #118898 as the feature tracking issue.

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
Yeah, I think this could use a lang-ok.

File: compiler/rustc_ast/src/ast.rs
Comment (compiler-errors) [inline compiler/rustc_ast/src/ast.rs:1429]: Don't use a bool here. Make an enum like `ForLoopKind`.

File: library/core/src/async_iter/mod.rs
Comment (compiler-errors) [inline library/core/src/async_iter/mod.rs:127]: does this even need to be publically exported?

**Review by compiler-errors — CHANGES_REQUESTED** [review]

**Comment by eholk** [comment]
Oh, I also need to add a test to make sure we don't parse `for await` without the feature.

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:1614]: This comment seems wrong

`// Iterator::next(&mut iter)`

then below on line 1624 it should be:

`// match $next_expr { .. }`

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:1598]: Please split this up across the several `let` exprs below for clarity of what each line is doing.

File: library/core/src/async_iter/mod.rs
Comment (eholk) [reply library/core/src/async_iter/mod.rs:127]: It looks like it does. When I took out the `pub` I got errors in the desugaring about accessing private types.

File: compiler/rustc_ast/src/ast.rs
Comment (eholk) [reply compiler/rustc_ast/src/ast.rs:1429]: Done

**Comment by the8472** [comment]
The proposed desugaring doesn't seem great. It'll have to poll a future and then match an option on each loop iteration, which means 2 conditions per iteration which is just terrible for optimizations. It's already hard to make regular for loop desugaring efficient and in some cases internal iteration still outperforms it. This would be much worse.

I think a pattern closer to nonblocking IO would be better where one keeps reading until EAGAIN and only then polls (awaits). That way things could be split into an inner next() + outer await loop.

Alternatively an async equivalent to `try_for_each` where all control flow is funneled through an enum might work too, though that'd require more work by the compiler.

**Comment by the8472** [comment]
Perhaps the signature of `async_iter_next` could be changed to return a `I: IntoIterator<IntoIter=IN>, IN: ExactSizeIterator`. `Option` fulfills those bounds. So single-item-per-await async iters could continue to return options while batchable async iters could return longer iterables.

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:1734]: You should not need to call `async_iter_next` at all, actually. I think you can just call `poll_next` in a loop like `await` does, perhaps even sharing logic. This doesn't necessarily block this PR, since I could also just implement this myself and it's an implementation detail at best.

Comment (eholk) [reply compiler/rustc_ast_lowering/src/expr.rs:1734]: I think it's cleaner not to have `async_iter_next` at all, so I'll go ahead and make that change in this PR. We're waiting on lang team approval anyway, so there's no rush.

**Comment by eholk** [comment]
Okay, I've reworked this PR significantly to get rid of `async_iter_next` and generate code that calls `poll_next` directly.

**Comment by lightsing** [comment]
Hi,

While I appreciate the efforts to introduce for await loops in Rust, I have some reservations about the proposed syntax and semantics.

In the current design, await is consistently used as a postfix operator (.await). Introducing for await breaks this consistency, potentially leading to confusion. The semantic clarity of await as a postfix operator has been a strong point in Rust's async design, and deviating from this pattern might weaken the language's overall coherence.

Personally, I would advocate for a syntax more in line with the existing async constructs. For example, `async for e in async_iter` seems more intuitive.

I believe this change warrants a more extensive discussion within the community. The introduction of a new syntax, especially one that deviates from established patterns, is a significant decision. It's crucial to consider the implications on the language's learning curve, consistency, and future extensibility.

While I support the initiative to enhance async iteration in Rust, I would urge reconsideration of the syntax to align better with existing conventions. A broader discussion might yield a more agreeable and consistent solution.

**Comment by asquared31415** [comment]
Note that in edition 2015 `for await in iter` is valid code, creating a binding with the identifier `await`, since it was not a reserved keyword at that time.  There should be syntax tests for that edition as well.

**Comment by slanterns** [comment]
It does not introduce a new `async` block but rather implicitly doing `await`-like thing in the desugaring, so I think `await` is probably more reasonable than `async`. Anyway it may not be necessary to bikeshedding on the syntax at this point :)

**Comment by traviscross** [comment]
> While I appreciate the efforts to introduce for await loops in Rust, I have some reservations about the proposed syntax and semantics.... this change warrants a more extensive discussion within the community.

Keep in mind that this is just a nightly experiment at this point.  No sticky decisions have been made on these details and they are still the subject of much reflection and discussion among the people working on this.

The purpose of the nightly experiment is to collect the kind of experience that can only come from implementing and trying the feature in nightly so that we can better write an RFC.

While feedback and participation is of course welcome here, the RFC process will be the main venue for the extensive discussion with the community that you mention.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

In the T-lang meeting, we agreed to proceed on this experiment.  @compiler-errors said we would land this under a separate feature gate.  @tmandry agreed to liaison.

Note that this is still just an experiment and will need an RFC after the details are worked out.

**Comment by eholk** [comment]
Thank you everyone for your feedback! That's definitely helpful and something we will consider as we continue developing this feature. As has been said a few times, the main goal is to get something in the compiler that we can start playing with and get a feel for. Sometimes it's hard to know what will feel right without being able to write and run programs. This is very much an experiment though and there will be lots more discussion before we are ready to stabilize anything. We'd love to have your continued input in the [#wg-async zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async).

The syntax as implemented in this PR is my preferred syntax at the moment, but I am open to changing my mind and adjusting the parser to use different syntaxes is an easy change to do.

We discussed this in the @rust-lang/wg-async planning meeting this morning. [Here are the minutes](https://hackmd.io/ZcJ3PGpWSLymr65NeyBKWA#Asynchronous-for-syntax). We had some good discussion that raised some points I hadn't considered.

I do see the point about moving `.await` to be prefix in this case seeming inconsistent. Overall, I think postfix await doesn't really make sense in a pattern for a couple reasons.

First, we are not actually awaiting the item. We are waiting for the iterator to give us an item. This is sort of the same as how a synchronous for loop hides the `Some(x)` and `None` that's returned from the `Iterator` trait's `next()` method. I feel like the analogy here would be if we were to write `for x.unwrap() in iter` in synchronous loops. We don't need to unwrap because the `for` loop does the match for us.

Second, I think of this in terms of introduction and elimination forms. In other words, this is the syntax you use to create an object or undo an object. For example, creating a reference is done with `&x` while you remove the reference by using `*x`. For structs, you create them by doing `Foo { x: 42 }`, but you access its fields by doing `foo.x`. In patterns, as far as I know, we always use the introduction form. We write `let Foo { x } = foo;` and not `let _.x = foo;`.

For async, `.await` is the elimination form, while `async { }` is the introduction form. So, to be consistent if we wanted to await futures in patterns, we'd do something like `let async { x } = future;` or `for async { x } in iter`. But I think this also brings us back to the fact that we don't have an iterator of futures, but an asynchronous iterator.

In my mind, an async for loop is a new kind of control flow construct, rather than a composition of several existing constructs, the same way that an async iterator is a new kind of object and not just an `Iterator<Item = impl Future>`. So I think it's best to think of this as distinct syntax for a distinct thing.

So this brings us to the question of how to spell an asynchronous for loop? One option is `async for x in async_iter`. I think this is reasonable in a lot of ways, but it's not my preferred syntax. The reason again goes back to intro and elimination forms. The `async` keyword is used to introduce or create a future, while `await` is used to complete a future. Asynchronous for loops are not creating futures though, they are completing them, so `await` seems like the right thing to do here. We could write `await for x in async_iter`, but I think `for await x in async_iter` reads better.

There's also a grepability argument in favor of using `await`, which is that right not every point at which an async function or block might suspend is marked with `await` (except for the implicit suspension point at the start of the function). This seems like a useful property to preserve. Seeing `for await` means that the function can suspend in the loop header.

Anyway, that's my perspective as a language designer/implementor and not so much as someone who writes a lot of async code. It might be that some design might make a lot of mathematical sense (from my perspective, anyway) but be completely surprising and non-intuitive to most users. I'd say it's more important for a language to make sense to its users than to satisfy some kind of language design ideals (those ideals should be in support of making something understandable!), so if this reasoning doesn't make sense then that's useful feedback and we should work towards something better.

File: compiler/rustc_ast/src/ast.rs
Comment (estebank) [inline compiler/rustc_ast/src/ast.rs:1429]: Instead of making the `await` part of the `for`, I think I would prefer we made it part of the `Pat` in the AST and reject it later in any pattern that isn't in a for loop. The current syntax implies that `let await foo = future;` would be valid. It isn't but we should likely handle it more gracefully. Particularly if there's *any* possibility that we'll allow people to write `for Struct { await foo } in stream` or `match foo { Some(await foo) => {}, None => {} }`.

Comment (compiler-errors) [reply compiler/rustc_ast/src/ast.rs:1429]: I think I disagree with this approach.

I don't think that we need to tangle up parsing `async PAT` in this PR. First of all, it makes lowering of `for await` in the HIR more difficult, since it makes the "is this a `for await` pattern" condition slightly more non-local (we need to peel off an async pat), and requires complicating the lowering of patterns with a new unreachable arms and/or error recovery.

Second, we discussed at length in a recent wg-async meeting that the idea of a `async PAT` syntax is not desirable at the moment. We also discussed at length whether we even *want* to call it `for await`, instead of something like `async for`, `for async`, etc. If we were to move to that, we'd have to undo all of the `PatKind::Async` here, too.

Last, if we later find that people are accidentally writing `let await PAT = EXPR;` enough that diagnostics should mention them specifically, then we can handle the recovery for that typo separately and without having to represent it in the AST. Simply eagerly erroring in `parse_pat_with_range_pat` if `may_recover && check_kw(kw::Await)` would be simple enough, I think.

Comment (compiler-errors) [reply compiler/rustc_ast/src/ast.rs:1429]: @eholk's comment above (https://github.com/rust-lang/rust/pull/118847#issuecomment-1857000155) discusses a bit about how `for await` shouldn't be conceptually thought of as `for (await PAT)` but as a totally separate kind of loop, and I tend to agree with that.

I do think that this PR should stay on the simple side -- I find that unnecessarily making parsing more general to carve out space for AST that doesn't exist is a good way to introduce new bugs, and make things like exhaustively matching on `PatKind`s that are only expected on an error path a bit more tedious.

Comment (compiler-errors) [reply compiler/rustc_ast/src/ast.rs:1429]: (also, I can also see some issues with pattern macro fragments either not matching `async PAT` correctly, or going from pass->fail if we're not careful, because currently `async` is not allowed in `nonterminal_may_begin_with` for pat fragments)

File: tests/ui/async-await/for-await.rs
Comment (compiler-errors) [inline tests/ui/async-await/for-await.rs:8]: Please add a test that this works correctly in an async-gen fn as well. Maybe just one that does:

```
for await i in iter { yield i; }
```

Or something to pass-through the awaited values.

File: tests/ui/typeck/issue-110052.stderr
Comment (compiler-errors) [inline tests/ui/typeck/issue-110052.stderr:12]: This is fine, though I think it would be desirable to filter out unstable associated items here... Not your fault, though, and can be done in another PR.

File: src/tools/rustfmt/src/expr.rs
Comment (compiler-errors) [inline src/tools/rustfmt/src/expr.rs:697]: Just pass through the `ForLoopKind`, this seems just as complicated but easier to get wrong 😆

File: src/tools/rustfmt/src/expr.rs
Comment (compiler-errors) [inline src/tools/rustfmt/src/expr.rs:793]: exhaustive match on `ForLoopKind` here

File: library/core/tests/async_iter/mod.rs
Comment (compiler-errors) [inline library/core/tests/async_iter/mod.rs:8]: This can use the `Waker::noop` API

File: library/core/src/async_iter/async_iter.rs
Comment (compiler-errors) [inline library/core/src/async_iter/async_iter.rs:141]: Do we need to introduce this now? I'd rather not need to get T-libs-api sign off for this... but I think this would need T-libs-api sign-off.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_ast_lowering/src/expr.rs
Comment (compiler-errors) [inline compiler/rustc_ast_lowering/src/expr.rs:62]:
```suggestion
                // correspond to the `e.id`, so `lower_expr_for` handles attribute lowering itself.
```

File: compiler/rustc_parse/src/parser/expr.rs
Comment (estebank) [inline compiler/rustc_parse/src/parser/expr.rs:2693]: Is there any reason in particular to not write this as follow with a single `if` check?

```suggestion
        let is_await =
            self.token.uninterpolated_span().at_least_rust_2018() && self.eat_keyword(kw::Await);
        let kind = if is_await {
            self.sess.gated_spans.gate(sym::async_for_loop, self.prev_token.span);
            ForLoopKind::ForAwait
        } else {
            ForLoopKind::For
        };
```

File: tests/ui/async-await/for-await.rs
Comment (eholk) [reply tests/ui/async-await/for-await.rs:8]: That turned out to be really helpful because I realized my desugaring wasn't pinning the iterator so to use it with `async gen fn` you'd have to use the `pin!` yourself, which seems pretty unfortunate.

File: library/core/src/async_iter/async_iter.rs
Comment (eholk) [reply library/core/src/async_iter/async_iter.rs:141]: I'll factor it out.

File: tests/ui/typeck/issue-110052.stderr
Comment (eholk) [reply tests/ui/typeck/issue-110052.stderr:12]: I think this should go away once I factor out the `IntoAsyncIterator` change.

**Comment by traviscross** [comment]
We discussed this in the WG-async triage meeting today.  The team was excited about this experiment.

**Comment by eholk** [comment]
> Note that in edition 2015 `for await in iter` is valid code, creating a binding with the identifier `await`, since it was not a reserved keyword at that time. There should be syntax tests for that edition as well.

Good point. I added `tests/ui/async-await/for-await-2015.rs` to make sure this works.

File: library/core/src/async_iter/async_iter.rs
Comment (eholk) [reply library/core/src/async_iter/async_iter.rs]: Okay, I've pulled out the `IntoAsyncIterator` changes and will push them to a second branch that I'll turn into PR and get T-libs-api signoff on.

File: tests/ui/async-await/for-await-consumes-iter.stderr
Comment (compiler-errors) [inline tests/ui/async-await/for-await-consumes-iter.stderr:22]: diagnostic bug but whatever, it's ok

**Review by compiler-errors — APPROVED** [review]
🆒

File: tests/ui/async-await/for-await-consumes-iter.stderr
Comment (estebank) [reply tests/ui/async-await/for-await-consumes-iter.stderr:22]: The span is marked with the appropriate reason, so it should be easy to fix later.

Comment (compiler-errors) [reply tests/ui/async-await/for-await-consumes-iter.stderr:22]: Yeah, it will also go away if/when we introduce `IntoAsyncIterator` which I had asked to separate from this PR so we can involve t-libs-api separately, so I don't see any reason to fix it here.

**Comment by compiler-errors** [comment]
please bless tests

then r=me

**Comment by compiler-errors** [comment]
No idea what's up with the test failure..

@bors retry

**Comment by apiraino** [comment]
since merge happened, I'll assume the I-async nomination can be now removed

@rustbot label -I-async-nominated

**Pull request merged**
